common/src/main/java/org/apache/comet/vector/CometVector.java

```diff
@@ -115,24 +121,35 @@ private Decimal createDecimal(BigDecimal value, int precision, int scale) {
     return dec;
   }
 
+  // bytes.length must be 16
+  public long getLongFromDecimalBytes(byte[] bytes) {
+    assert (bytes.length == 16);
+    // get Long value from the last eight bytes
+    // Use ByteBuffer's fast conversion to long
+    long val = ByteBuffer.wrap(bytes).getLong(8);
+    return val;
+  }
+
   /**
    * Reads a 16-byte byte array which are encoded big-endian for decimal128 into internal byte
    * array.
    */
   byte[] getBinaryDecimal(int i) {
+    // TODO: consider implementing a zero-copy version
     return copyBinaryDecimal(i, DECIMAL_BYTES);
   }
 
   /** Reads a 16-byte byte array which are encoded big-endian for decimal128. */
   public byte[] copyBinaryDecimal(int i, byte[] dest) {
-    long valueBufferAddress = getValueVector().getDataBuffer().memoryAddress();
-    Platform.copyMemory(
-        null,
-        valueBufferAddress + (long) i * DECIMAL_BYTE_WIDTH,
-        dest,
-        Platform.BYTE_ARRAY_OFFSET,
-        DECIMAL_BYTE_WIDTH);
+    ValueVector vector = getValueVector();
+    // If the index is zero and DECIMAL_BYTES_ALL already has data, we have a new
+    // batch of data in the vector's backing buffer. So read it again.
```

Comment:
Could you expand on why `i==0` means that have new data? I'm not entirely clear on why this is the case.

Reply:
This check is now unnecessary so I've removed it. For reference, `ColumnReader.loadVector` specified the rules under which the vector may be reused for a new batch of data. When a new batch is read into the same vector, the underlying memory buffer is reused and there is no change in the buffer memory address of the value vector. The check for `index==0` banked on the fact that the iterator reading the values will start from the beginning every time a new batch is read. There was a gotcha though. If the first N values were null the iterator would skip this method and call it for the first time with an index of N which could be non-zero.
The only safe way is to reset the buffer when a new batch is read (which is what this PR does now).